src/meltano/core/state_service.py

```diff
@@ -184,3 +184,26 @@ def merge_state(self, state_id_src: str, state_id_dst: str):
         src_state_dict = self.get_state(state_id_src)
         src_state = json.dumps(src_state_dict)
         self.add_state(state_id_dst, src_state, payload_flags=Payload.INCOMPLETE_STATE)
+
+    def copy_state(self, state_id_src: str, state_id_dst: str):
+        """Copy state from Job state_id_src onto Job state_id_dst.
+
+        Args:
+            state_id_src: the state_id to get state from
+            state_id_dst: the state_id_to copy state onto
+        """
+        src_state_dict = self.get_state(state_id_src)
+        src_state = json.dumps(src_state_dict)
+        self.set_state(state_id_dst, src_state)
+
+    def move_state(self, state_id_src: str, state_id_dst: str):
+        """Move state from Job state_id_src to Job state_id_dst.
+
+        Args:
+            state_id_src: the state_id to get state from and clear
+            state_id_dst: the state_id_to move state onto
+        """
+        src_state_dict = self.get_state(state_id_src)
+        src_state = json.dumps(src_state_dict)
+        self.set_state(state_id_dst, src_state)
+        self.clear_state(state_id_src)
```

Comment:
(From @edgarrmondragon's gitlab review comment)
> of the four lines of code here, what happens if 4 fails for any reason? Does it mean state is set at the destination ID but the src ID is not removed?
Is there a way to rollback 3 if 4 fails?
Surely a nitpick because if 4 fails, it's as easy as deleting the lingering ID.

I agree that it's a concern, not sure how to address.

Here's my quick attempt at adding a rollback:
```py
    def move_state(self, state_id_src: str, state_id_dst: str):
        """Move state from Job state_id_src to Job state_id_dst.

        Args:
            state_id_src: the state_id to get state from and clear
            state_id_dst: the state_id_to move state onto
        """
        # Preserve starting states
        src_state_dict = self.get_state(state_id_src)
        src_state = json.dumps(src_state_dict)
        dst_state_dict = self.get_state(state_id_dst)
        dst_state = json.dumps(dst_state_dict)
        try:
            # Attempt the operation
            self.set_state(state_id_dst, src_state)
            self.clear_state(state_id_src)
        except:
            # Rollback both state-ids and raise the caught exception
            self.set_state(state_id_src, src_state)
            self.set_state(state_id_dst, dst_state)
            raise
```
At first glance it seems a bit overly defensive, especially since a bunch of the methods on this service rely on other common methods. The `clear_state` function is implemented via `clear` -> `set` -> `add` just with an empty payload. So if there was a problem in the `clear_state` method, it could also come up again when we attempt to rollback with `set`.

But that's just my attempt. Perhaps someone who better understands error handling patterns in python could weigh in. There might be an opportunity to make the whole class handle exceptions better.

Reply:
@rwfeather Would using a transaction work? e.g.

```python
        with self.session.begin():
            self.set_state(state_id_dst, src_state)
            self.clear_state(state_id_src)
```

Reply:
@edgarrmondragon @rwfeather - the `add_state` method that gets called in the call stack for these methods calls `Job.save(self.session)` which commits the session, so I'm not sure that you could wrap this whole thing in a transaction as-is, you would probably need to add a flag to that method indicating whether or not to commit

Reply:
@cjohnhanson Oh, then that'd be trickier that I had hoped. I'm OK with leaving this out then 👍

Issue for the future: https://github.com/meltano/meltano/issues/6279